lib/communicator_mpi.cpp

```diff
@@ -0,0 +1,355 @@
+{
+  MsgHandle *mh = (MsgHandle *)safe_malloc(sizeof(MsgHandle));
+  MPI_CHECK(MPI_Send_init(buffer, nbytes, MPI_BYTE, rank, tag, MPI_COMM_HANDLE, &(mh->request)));
+  mh->custom = false;
+
+  return mh;
+}
+
+/**
+ * Declare a message handle for receiving `nbytes` from the `rank` with `tag`.
+ */
+MsgHandle *Communicator::comm_declare_recv_rank(void *buffer, int rank, int tag, size_t nbytes)
+{
+  MsgHandle *mh = (MsgHandle *)safe_malloc(sizeof(MsgHandle));
+  MPI_CHECK(MPI_Recv_init(buffer, nbytes, MPI_BYTE, rank, tag, MPI_COMM_HANDLE, &(mh->request)));
+  mh->custom = false;
+
+  return mh;
+}
+
+/**
+ * Declare a message handle for sending to a node displaced in (x,y,z,t) according to "displacement"
+ */
+MsgHandle *Communicator::comm_declare_send_displaced(void *buffer, const int displacement[], size_t nbytes)
+{
+  Topology *topo = comm_default_topology();
+  int ndim = comm_ndim(topo);
+  check_displacement(displacement, ndim);
+
+  int rank = comm_rank_displaced(topo, displacement);
+
+  int tag = 0;
+  for (int i = ndim - 1; i >= 0; i--) tag = tag * 4 * max_displacement + displacement[i] + max_displacement;
+  tag = tag >= 0 ? tag : 2 * pow(4 * max_displacement, ndim) + tag;
+
+  MsgHandle *mh = (MsgHandle *)safe_malloc(sizeof(MsgHandle));
+  MPI_CHECK(MPI_Send_init(buffer, nbytes, MPI_BYTE, rank, tag, MPI_COMM_HANDLE, &(mh->request)));
+  mh->custom = false;
+
+  return mh;
+}
+
+/**
+ * Declare a message handle for receiving from a node displaced in (x,y,z,t) according to "displacement"
+ */
+MsgHandle *Communicator::comm_declare_receive_displaced(void *buffer, const int displacement[], size_t nbytes)
+{
+  Topology *topo = comm_default_topology();
+  int ndim = comm_ndim(topo);
+  check_displacement(displacement, ndim);
+
+  int rank = comm_rank_displaced(topo, displacement);
+
+  int tag = 0;
+  for (int i = ndim - 1; i >= 0; i--) tag = tag * 4 * max_displacement - displacement[i] + max_displacement;
+  tag = tag >= 0 ? tag : 2 * pow(4 * max_displacement, ndim) + tag;
+
+  MsgHandle *mh = (MsgHandle *)safe_malloc(sizeof(MsgHandle));
+  MPI_CHECK(MPI_Recv_init(buffer, nbytes, MPI_BYTE, rank, tag, MPI_COMM_HANDLE, &(mh->request)));
+  mh->custom = false;
+
+  return mh;
+}
+
+/**
+ * Declare a message handle for sending to a node displaced in (x,y,z,t) according to "displacement"
+ */
+MsgHandle *Communicator::comm_declare_strided_send_displaced(void *buffer, const int displacement[], size_t blksize,
+                                                             int nblocks, size_t stride)
+{
+  Topology *topo = comm_default_topology();
+  int ndim = comm_ndim(topo);
+  check_displacement(displacement, ndim);
+
+  int rank = comm_rank_displaced(topo, displacement);
+
+  int tag = 0;
+  for (int i = ndim - 1; i >= 0; i--) tag = tag * 4 * max_displacement + displacement[i] + max_displacement;
+  tag = tag >= 0 ? tag : 2 * pow(4 * max_displacement, ndim) + tag;
```

Comment:
I don't necessarily like the use of `pow` here, but I'm also not sure why the compiler isn't complaining about the conversion to `double`... is there an `int` power defined somewhere? I don't see one here:

http://www.cplusplus.com/reference/cmath/pow/

Reply:
This is used in the original code. For C++11 there is an official overloading - click the C++11 on that page and you will see it. That's probably why there is no compiler warning. Basically `int`s are converted to `double`s, power, and convert back.

Reply:
Yes, they template on the input type, not the output type? Unless there's a typo on the page.

Reply:
Anyway, if it's not barfing on it, I'll let it be.